rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/custom_participant_info.hpp

```diff
@@ -15,19 +15,82 @@
 #ifndef RMW_FASTRTPS_CPP__CUSTOM_PARTICIPANT_INFO_HPP_
 #define RMW_FASTRTPS_CPP__CUSTOM_PARTICIPANT_INFO_HPP_
 
+#include <map>
+#include <string>
+#include <vector>
+
+#include "fastrtps/attributes/ParticipantAttributes.h"
 #include "fastrtps/participant/Participant.h"
+#include "fastrtps/participant/ParticipantListener.h"
 
+#include "rmw/impl/cpp/key_value.hpp"
 #include "rmw/rmw.h"
 
 #include "rmw_fastrtps_cpp/reader_info.hpp"
 #include "rmw_fastrtps_cpp/writer_info.hpp"
 
+class ParticipantListener;
+
 typedef struct CustomParticipantInfo
 {
   eprosima::fastrtps::Participant * participant;
+  ::ParticipantListener * listener;
   ReaderInfo * secondarySubListener;
   WriterInfo * secondaryPubListener;
   rmw_guard_condition_t * graph_guard_condition;
 } CustomParticipantInfo;
 
+class ParticipantListener : public eprosima::fastrtps::ParticipantListener
+{
+public:
+  void onParticipantDiscovery(Participant *, ParticipantDiscoveryInfo info) override
+  {
+    if (
+      info.rtps.m_status != DISCOVERED_RTPSPARTICIPANT &&
+      info.rtps.m_status != REMOVED_RTPSPARTICIPANT &&
+      info.rtps.m_status != DROPPED_RTPSPARTICIPANT)
+    {
+      return;
+    }
+
+    if (DISCOVERED_RTPSPARTICIPANT == info.rtps.m_status) {
+      // ignore already known GUIDs
+      if (discovered_names.find(info.rtps.m_guid) == discovered_names.end()) {
+        auto map = rmw::impl::cpp::parse_key_value(info.rtps.m_userData);
+        auto found = map.find("name");
+        std::string name;
+        if (found != map.end()) {
+          name = std::string(found->second.begin(), found->second.end());
+        }
+        if (name.empty()) {
+          // use participant name if no name was found in the user data
+          name = info.rtps.m_RTPSParticipantName;
+        }
+        // ignore discovered participants without a name
+        if (!name.empty()) {
+          discovered_names[info.rtps.m_guid] = name;
+        }
+      }
+    } else {
+      auto it = discovered_names.find(info.rtps.m_guid);
+      // only consider known guid's
```

Comment:
guiid's -> GUIDs

Reply:
I only changed the line you commented on before and didn't see the second occurrence. This one is changed too now. (Please feel free to apply changes like this directly.)